compiler/scenario-service/client/src/DA/Daml/LF/ScenarioServiceClient/LowLevel.hs

```diff
@@ -395,14 +403,94 @@ performRequest method timeoutSeconds payload = do
     ClientNormalResponse _ _ _ status _ -> return (Left $ BErrorFail status)
     ClientErrorResponse err -> return (Left $ BErrorClient err)
 
+runBiDiLive
+  :: (SS.ScenarioService ClientRequest ClientResult
+      -> ClientRequest 'BiDiStreaming SS.RunScenarioRequest SS.RunScenarioResponseOrStatus
+      -> IO (ClientResult 'BiDiStreaming SS.RunScenarioResponseOrStatus))
+  -> Handle -> ContextId -> LF.ValueRef -> Logger -> MVar Bool -> (SS.ScenarioStatus -> IO ())
+  -> IO (Either Error SS.ScenarioResult)
+runBiDiLive runner Handle{..} (ContextId ctxId) name logger stopSemaphore statusUpdateHandler = do
+  let startReq =
+        SS.RunScenarioRequest $ Just $ SS.RunScenarioRequestSumStart $
+          SS.RunScenarioStart ctxId (Just (toIdentifier name))
+  let cancelReq = SS.RunScenarioRequest $ Just $ SS.RunScenarioRequestSumCancel SS.RunScenarioCancel
+
+  (setFinalResponse, getFinalResponse) <- do
+    -- Hide finalResponse inside closure
+    finalResponse <- newIORef Nothing
+    let set r =
+          atomicModifyIORef finalResponse $ \old ->
+            let err =
+                  case old of
+                    Just (Left _) -> old
+                    _ -> Just r
```

Comment:
The combination of the Maybe and Either to represent no response, error or success might be better handled by its own sum?
if you're daring, you could even give it the reasonable semigroup behaviour to change this instance into `<>` plus some fluff
Curious what you think (for this and future cases)

Reply:
It seems pretty sophisticated for what essentially amounts to glue code
But, I think it can't hurt readability of the code, and I do love me a semigroup :)